### PR TITLE
Implement context handling in requests

### PR DESCRIPTION
**Pull request by defelmnq**
Implementation of a timeout in our HTTP requests and a retries policy.

### CODE REVIEW

File: gogoogledm.go [hunk 1]
```diff
@@ -158,10 +159,22 @@ func (api *DistanceMatrixAPI) sendRequest(origins []Coordinates, destinations []
...
 	}
 
-	resp, err := http.Get(url)
+	client := http.Client{
+		Timeout: 500 * time.Millisecond,
+	}
+
+	var resp *http.Response
+
+	for retries := 0; retries < maxRetries; retries++ {
```
Comment (christophe-dufour) [inline gogoogledm.go]: As far as I get the code, you will retry even if there is no error...

Comment (defelmnq) [reply gogoogledm.go]: Fixed, missed a commit here 🙈

File: gogoogledm.go [hunk 2]
```diff
@@ -147,7 +149,7 @@ func (api *DistanceMatrixAPI) generateAuthentifiedURL(urlValues url.Values) (str
 	return (base_host + base_path + signedQuery), nil
 }
 
-func (api *DistanceMatrixAPI) sendRequest(origins []Coordinates, destinations []Coordinates, transportMode TransportMode) (*ApiResponse, error) {
+func (api *DistanceMatrixAPI) sendRequest(ctx context.Context, origins []Coordinates, destinations []Coordinates, transportMode TransportMode) (*ApiResponse, error) {
```
Comment (christophe-dufour) [inline gogoogledm.go:151]: :+1:

File: example/main.go [hunk 3]
```diff
@@ -2,7 +2,11 @@ package main
 
 import (
 	"fmt"
-	. "github.com/jondunning/gogoogledm"
+	"time"
```
Comment (christophe-dufour) [inline example/main.go]: I guess this will be removed?

File: gogoogledm.go [hunk 4]
```diff
@@ -9,6 +9,8 @@ import (
 	"net/url"
 	"strings"
 	"time"
+
+	"golang.org/x/net/context"
```
Comment (christophe-dufour) [inline gogoogledm.go]: I think it should be better to use "context" here. But no strong opinion.

@arkan WDYT?

Comment (defelmnq) [reply gogoogledm.go]: Same for me, not having a strong opinion on that point and can change if we prefer to have "context".

File: example/main.go (hunk 3 above)
Comment (defelmnq) [inline example/main.go:5]: the `time` ? It is used by the package for the field `timeToWait` (was already there).

**Comment by defelmnq** [comment]
🆙 - the package has been tested and looks ready !

**Comment by asdine** [comment]
🌷 If i'm not wrong, this fork is unecessary. You just need to replace the http default client with a new one with a timeout and you're done.

```go
http.DefaultClient = &http.Client{
	Timeout: 10 * time.Second,
}
```

**Comment by defelmnq** [comment]
Answered more explicitly in the universe's PR but yes, we explored this solution but want two things : 
- avoid as much as possible having to work on the library itself (excepted here to include context, something generic)
- having easily configurable timeout and be able to cumulate timeouts

**Review by christophe-dufour — COMMENTED** [review]
Just change to import to "context" and it's 👍

**Comment by defelmnq** [comment]
🆙 with context

**Review by christophe-dufour — APPROVED** [review]

**Comment by arkan** [comment]
@Nyks06 can you squash your commits? 

EDIT: and rename the title of the PR?

**Review by genesor — APPROVED** [review]

**Pull request merged**